src/analysis/calculus/uniform_limits_deriv.lean

```diff
@@ -0,0 +1,394 @@
+/-
+Copyright (c) 2022 Kevin H. Wilson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin H. Wilson
+-/
+import analysis.calculus.fderiv
+import analysis.calculus.mean_value
+import analysis.convex.topology
+import analysis.normed_space.is_R_or_C
+import data.complex.is_R_or_C
+import order.filter.basic
+
+/-!
+# Swapping limits and derivatives via uniform convergence
+
+The purpose of this file is to prove that the derivative of the pointwise limit of a sequence of
+functions is the pointwise limit of the functions' derivatives when the derivatives converge
+_uniformly_. The formal statement appears as `has_fderiv_at_of_tendsto_uniformly_on`.
+
+## Main statements
+
+* `has_fderiv_at_of_tendsto_uniformly_on` : If `f : ℕ → E → G` is a sequence of functions with
+  derivatives `f' : ℕ → (E → (E →L[𝕜] G))` and the `f` converge pointwise to `g` and the `f'`
+  converge _uniformly_ on some closed ball, then the derivative of `g'` is the pointwise limit
+  of the `f'` on the closed ball
+
+## Implementation notes
+
+Our proof utilizes three major components:
+  * `convex.norm_image_sub_le_of_norm_has_fderiv_within_le`: The mean value inequality for
+    vector-valued functions over `ℝ` and `ℂ`
+  * `norm_add_le`: The triangle inequality
+  * `uniform_cauchy_seq_on.tendsto_uniformly_on_of_tendsto` which allows us to upgrade pointwise
+    convergence to uniform convergence by showing that the Cauchy sequences converge uniformly to 0
+
+## Tags
+
+uniform convergence, limits of derivatives
+-/
+
+open finset filter metric
+open_locale uniformity filter
+
+section convenience_lemmas
+
+/-- Convenience lemma -/
+private lemma mul_cancel_inv_left₀ {a b : ℝ} (ha : a ≠ 0) : a⁻¹ * (a * b) = b :=
+ begin
+  conv { congr, congr, skip, rw ←inv_inv a, },
+  have : a⁻¹ ≠ 0, simp [ha],
+  rw mul_inv_cancel_left₀ this,
+end
+
+/-- We'll need to apply the triangle ineqaulity twice, which is easier to do in lemma
+format -/
+private lemma norm_add_three_le {E : Type*} [semi_normed_group E] {a b c : E} :
+  ∥a + b + c∥ ≤ ∥a∥ + ∥b∥ + ∥c∥ :=
+begin
+  refine le_trans (norm_add_le _ _) _,
+  exact add_le_add (norm_add_le _ _) rfl.le,
+end
+
+/-- Convenience lemma for three way comparisons -/
+private lemma lt_of_lt_of_lt_of_le {a b c a' b' c': ℝ} :
+  a < a' → b < b' → c ≤ c' → a + b + c < a' + b' + c' :=
+begin
+  intros ha hb hc,
+  linarith [ha, hb, hc],
+end
+
+end convenience_lemmas
+
+section limits_of_derivatives
+
+variables {E : Type*} [normed_group E] {𝕜 : Type*} {G : Type*}
+  [is_R_or_C 𝕜] [normed_group G] [normed_space 𝕜 G]
+  {f : ℕ → E → G} {g : E → G} {x y z : E} {r R C : ℝ}
+
+/-- For fixed y and z, if `f_n → g` pointwise, then the difference quotients
+`∥z - y∥⁻¹ • (f_n z - f_n y)` converge to the difference quotients
+`∥z - y∥⁻¹ • (g z - g y)` -/
+lemma difference_quotients_converge
+  (hfg : ∀ (y : E), y ∈ closed_ball x r → tendsto (λ n, f n y) at_top (nhds (g y))) :
```

Comment:
Is this really worth to be a separate lemma? Can you inline the proof?

Reply:
It is used a few times in the proof, so my software engineering instinct was to pull it out (and make it private if it's cluttering the API). Should I just inline the proof each time?